File: server/datamodel/payment.py
import os

import stripe


class Payment:
    def __init__(self, number="", exp_month=0, exp_year=0, cvc=0):
        self.price_id = os.getenv("STRIPE_PRICE_ID")
        resp = create_payment_method(number, exp_month, exp_year, cvc)
        self.method_id = resp
        resp = create_customer(self.method_id)
        self.customer_id = resp
        update_payment_method(self.method_id, self.customer_id)
        self.subscription_id = create_subscription(self.customer_id, self.price_id)


def create_payment_method(number="", exp_month=0, exp_year=0, cvc=0):
    resp = stripe.PaymentMethod.create(
        api_key=os.getenv("STRIPE_API_KEY"),
        type="card",
        card={
            "number": number,
            "exp_month": exp_month,
            "exp_year": exp_year,
            "cvc": cvc
        },
    )
    return resp['id']


def create_customer(method_id):
    resp = stripe.Customer.create(
        api_key=os.getenv("STRIPE_API_KEY"),
        payment_method=method_id,
        invoice_settings={
            "default_payment_method": method_id
        }
    )
    return resp['id']


def update_payment_method(method_id, customer_id):
    stripe.PaymentMethod.attach(
        method_id,
        api_key=os.getenv("STRIPE_API_KEY"),
        customer=customer_id
    )


def create_subscription(customer_id, price_id):
    resp = stripe.Subscription.create(
        api_key=os.getenv("STRIPE_API_KEY"),
        customer=customer_id,
        items=[
            {"price": price_id},
        ],
    )
    return resp['id']


class Store:
    def __init__(self):
        super().__init__()
        self.payments = {}

    def add_payment(self, data):
        number = data['number']
        exp_month = data['exp_month']
        exp_year = data['exp_year']
        cvc = data['cvc']
        payment = Payment(number, exp_month, exp_year, cvc)
        self.payments[payment.subscription_id] = payment

    def change_status(self, subscription_id, status):
        self.payments[subscription_id].subscription_state = status

File: payment/payment.go
package payment

import (
	"github.com/google/uuid"
	"github.com/mkieweg/stripe-api/config"
	"github.com/stripe/stripe-go/v72"
	"github.com/stripe/stripe-go/v72/customer"
	"github.com/stripe/stripe-go/v72/paymentmethod"
	"github.com/stripe/stripe-go/v72/sub"
)

func NewPayment(req *PaymentRequest) (*Payment, error) {
	p := &Payment{
		PriceID: config.PriceID(),
		UUID:    uuid.New(),
	}

	pmid, err := createPaymentMethod(req)
	if err != nil {
		return nil, err
	}
	p.MethodID = *pmid

	cid, err := createCustomer(pmid)
	if err != nil {
		return nil, err
	}
	p.CustomerID = *cid
	return p, nil
}

type Payment struct {
	PriceID        string    `json:"priceId"`
	MethodID       string    `json:"paymentMethodId"`
	CustomerID     string    `json:"customerId"`
	SubscriptionID string    `json:"subscriptionId"`
	UUID           uuid.UUID `json:"uuid"`
}

func createPaymentMethod(req *PaymentRequest) (*string, error) {
	params := &stripe.PaymentMethodParams{
		Card: &stripe.PaymentMethodCardParams{
			Number:   &req.Number,
			ExpMonth: &req.ExpMonth,
			ExpYear:  &req.ExpYear,
			CVC:      &req.Cvc,
		},
		Type: stripe.String("card"),
	}
	resp, err := paymentmethod.New(params)
	if err != nil {
		return nil, err
	}
	return &resp.ID, nil
}

func createCustomer(id *string) (*string, error) {
	params := &stripe.CustomerParams{
		PaymentMethod: id,
		InvoiceSettings: &stripe.CustomerInvoiceSettingsParams{
			DefaultPaymentMethod: id,
		},
	}
	resp, err := customer.New(params)
	if err != nil {
		return nil, err
	}
	return &resp.ID, nil
}

func (p *Payment) AttachCustomer(customerID string) error {
	params := &stripe.PaymentMethodAttachParams{
		Customer: stripe.String(customerID),
	}
	_, err := paymentmethod.Attach(p.MethodID, params)
	return err
}

func (p *Payment) CreateSubscription() (*string, error) {
	params := &stripe.SubscriptionParams{
		Customer: stripe.String(p.CustomerID),
		Items: []*stripe.SubscriptionItemsParams{
			{
				Plan: stripe.String(p.PriceID),
			},
		},
	}
	params.AddExpand("latest_invoice.payment_intent")
	resp, err := sub.New(params)
	if err != nil {
		return nil, err
	}
	return &resp.ID, nil
}

type PaymentRequest struct {
	Number   string `json:"number"`
	ExpMonth string `json:"exp_month"`
	ExpYear  string `json:"exp_year"`
	Cvc      string `json:"cvc"`
}
